# This notebook is generated by Liucheng on Jul 10, 2018

# Question 1a: Load the data in the SQLite database into a **Pandas** dataframe. The dataframe should be indexed by unique keywords (48 rows), and the columns should correspond to the different weeks (261 columns). 

In [ ]:
import pandas as pd
import sqlite3
!ls

PART1.md             Part1_solution.ipynb denim_search_data.db


In [ ]:
# load the data from the SQLite database into a Pandas dataframe
with sqlite3.connect('denim_search_data.db') as con:
    df=pd.read_sql("select * from search_vol",con,index_col='keyword',parse_dates=['w0','w1'])

In [ ]:
print(df.shape)
print(df.dtypes)
print(df.head())
print(df.describe())

(11437, 3)
w0        datetime64[ns]
w1        datetime64[ns]
volume            object
dtype: object
                         w0         w1 volume
keyword                                      
ripped jeans     2013-02-03 2013-02-09  10887
relaxed jeans    2013-02-03 2013-02-09    381
wide leg jeans   2013-02-03 2013-02-09   5413
distressed jeans 2013-02-03 2013-02-09   6570
destroyed jeans  2013-02-03 2013-02-09   3554
                         w0                   w1 volume
count                 11437                11437  11437
unique                  261                  261   2620
top     2016-06-12 00:00:00  2017-03-18 00:00:00   2237
freq                     47                   47     40
first   2013-02-03 00:00:00  2013-02-09 00:00:00    NaN
last    2018-01-28 00:00:00  2018-02-03 00:00:00    NaN


In [ ]:
# rename the column, change the tpye of the column
df=df.rename(columns={'w0':'week_begin_date'})
df['volume']=df['volume'].astype('int64')
print(df.dtypes)

week_begin_date    datetime64[ns]
w1                 datetime64[ns]
volume                      int64
dtype: object


In [ ]:
# creat a pivot table to answer the question
df2=pd.pivot_table(df,values='volume',index=['keyword'],columns=['week_begin_date'])

In [ ]:
print(df2.shape)
df2.head()

(47, 261)


week_begin_date    2013-02-03  2013-02-10  2013-02-17  2013-02-24  2013-03-03  \
keyword                                                                         
ankle jeans            1434.0      1775.0      2321.0      1980.0      1912.0   
bell bottom jeans      4049.0      4526.0      3335.0      6669.0      6907.0   
bell bottoms           5760.0      5120.0      6080.0      6080.0      7040.0   
bootcut jeans         21832.0     24950.0     27624.0     26733.0     26733.0   
boyfriend jeans       31500.0     33923.0     42403.0     38769.0     43615.0   

week_begin_date    2013-03-10  2013-03-17  2013-03-24  2013-03-31  2013-04-07  \
keyword                                                                         
ankle jeans            2253.0      1912.0      2253.0      2321.0      1775.0   
bell bottom jeans      5955.0      4287.0      4049.0      5002.0      5240.0   
bell bottoms           8960.0      6720.0      5760.0      7680.0      5760.0   
bootcut jeans         28960.0     27624.0     29851.0     21386.0     23168.0   
boyfriend jeans       38769.0     38769.0     42403.0     52095.0     47249.0   

week_begin_date       ...      2017-11-26  2017-12-03  2017-12-10  2017-12-17  \
keyword               ...                                                       
ankle jeans           ...          6759.0      5735.0      5189.0      5257.0   
bell bottom jeans     ...         19055.0     16673.0     12624.0     14768.0   
bell bottoms          ...         17280.0     17920.0     22080.0     13120.0   
bootcut jeans         ...         41436.0     38317.0     40990.0     35198.0   
boyfriend jeans       ...         71480.0     67845.0     75114.0     60576.0   

week_begin_date    2017-12-24  2017-12-31  2018-01-07  2018-01-14  2018-01-21  \
keyword                                                                         
ankle jeans            5325.0      5667.0      4438.0      5598.0      5462.0   
bell bottom jeans     16435.0     15959.0     15244.0     14530.0     15006.0   
bell bottoms          19520.0     17280.0     20800.0     20480.0     24000.0   
bootcut jeans         40545.0     34307.0     32970.0     36980.0     33861.0   
boyfriend jeans       83595.0     71480.0     67845.0     81172.0     79960.0   

week_begin_date    2018-01-28  
keyword                        
ankle jeans            4847.0  
bell bottom jeans     15006.0  
bell bottoms          18240.0  
bootcut jeans         32970.0  
boyfriend jeans       83595.0  

[5 rows x 261 columns]

# My answer:

So the dataframe df2 is my answer that includes weekly searches for different styles of denim. It has 47 rows that correpond to 47 different styles of denim (not 48 styles of denim due to no search record for one specific style of denim). Its 261 columns correspond to the different weeks. 

# Question 1b: Based on the given data, which three denim styles do you believe will be the top performers of 2018? Which three will be the bottom performers?

In [ ]:
# For the performers of 2018, I took the data in the week from 2017-12-31 to 2018-01-06 into account.
df3=df[df['week_begin_date']>=pd.datetime(2017,12,31)]
df_year2018=pd.pivot_table(df3,values='volume',index=['keyword'],columns=['week_begin_date'])

In [ ]:
print(df_year2018.shape)
print(df_year2018.columns)

(47, 5)
DatetimeIndex(['2017-12-31', '2018-01-07', '2018-01-14', '2018-01-21',
               '2018-01-28'],
              dtype='datetime64[ns]', name=u'week_begin_date', freq=None)


In [ ]:
# top/bottom 3 performers of 2018
df_year2018['sum_2018']=df_year2018.sum(axis=1)
print('=== top 3 performers of 2018 ===')
print(df_year2018[['sum_2018']].nlargest(3,'sum_2018'))
print
print('=== bottom 3 performers of 2018 ===')
print(df_year2018[['sum_2018']].nsmallest(3,'sum_2018'))

=== top 3 performers of 2018 ===
week_begin_date   sum_2018
keyword                   
mom jeans        1298773.0
cropped jeans    1037067.0
jeggings          548335.0

=== bottom 3 performers of 2018 ===
week_begin_date          sum_2018
keyword                          
cropped frayed jeans        675.0
cropped boyfriend jeans     999.0
sculpt jeans               1119.0


# My answer:

For the performers of 2018, top 3 weekly searches are mom jeans (1298773 times), cropped jeans (1037067 times) and jeggings (548335 times). The bottom 3 weekly searches are cropped frayed jeans (675 times), cropped boyfriend jeans (999 times) and sculpt jeans (1119 times). Again, there is no search record for one specific style of denim in the database. I don't even know the name of denim that has no search record.

# Cross-check my answer:

In [ ]:
df_year2018.loc[['mom jeans','cropped jeans','jeggings']]

week_begin_date  2017-12-31 00:00:00  2018-01-07 00:00:00  \
keyword                                                     
mom jeans                    71657.0              72872.0   
cropped jeans                 9075.0             999999.0   
jeggings                    116247.0             116247.0   

week_begin_date  2018-01-14 00:00:00  2018-01-21 00:00:00  \
keyword                                                     
mom jeans                    82588.0              71657.0   
cropped jeans                 8767.0               9690.0   
jeggings                    105280.0             103087.0   

week_begin_date  2018-01-28 00:00:00   sum_2018  
keyword                                          
mom jeans                   999999.0  1298773.0  
cropped jeans                 9536.0  1037067.0  
jeggings                    107474.0   548335.0

In [ ]:
df_year2018.loc[['cropped frayed jeans','cropped boyfriend jeans','sculpt jeans']]

week_begin_date          2017-12-31 00:00:00  2018-01-07 00:00:00  \
keyword                                                             
cropped frayed jeans                     NaN                  NaN   
cropped boyfriend jeans                253.0                120.0   
sculpt jeans                           294.0                165.0   

week_begin_date          2018-01-14 00:00:00  2018-01-21 00:00:00  \
keyword                                                             
cropped frayed jeans                   407.0                268.0   
cropped boyfriend jeans                120.0                349.0   
sculpt jeans                           229.0                202.0   

week_begin_date          2018-01-28 00:00:00  sum_2018  
keyword                                                 
cropped frayed jeans                     NaN     675.0  
cropped boyfriend jeans                157.0     999.0  
sculpt jeans                           229.0    1119.0